src/ecdsapod.rs

```diff
@@ -0,0 +1,546 @@
+        // computed at further stages of the test (affecting the time reports)
+        timed!(
+            "generate ECDSA_VERIFY, STANDARD_ECDSA_POD_DATA, STANDARD_REC_MAIN_POD_CIRCUIT",
+            {
+                let (_, _) = &*ECDSA_VERIFY_DATA;
+                let (_, _) = &*STANDARD_ECDSA_POD_DATA;
+                let _ = &*pod2::backends::plonky2::STANDARD_REC_MAIN_POD_CIRCUIT_DATA;
+            }
+        );
+
+        let params = Params {
+            max_input_signed_pods: 0,
+            ..Default::default()
+        };
+
+        let sk = ECDSASecretKey::<Secp256K1>(Secp256K1Scalar::rand());
+        let msg = Secp256K1Scalar::rand();
+        // compute the pk & signature
+        let pk: ECDSAPublicKey<Secp256K1> =
+            ECDSAPublicKey((CurveScalar(sk.0) * Secp256K1::GENERATOR_PROJECTIVE).to_affine());
+        let signature: ECDSASignature<Secp256K1> = sign_message(msg, sk);
+
+        let vds_root = EMPTY_HASH;
+        let ecdsa_pod = timed!(
+            "EcdsaPod::new",
+            EcdsaPod::new(&params, vds_root, msg, pk, signature).unwrap()
+        );
+
+        ecdsa_pod.verify().unwrap();
+        // pod2::measure_gates_print!();
+
+        // wrap the ecdsa_pod in a 'MainPod' (RecursivePod)
+        let main_ecdsa_pod = pod2::frontend::MainPod {
+            pod: ecdsa_pod.clone(),
+            public_statements: ecdsa_pod.pub_statements(),
+            params: params.clone(),
+        };
+
+        // now generate a new MainPod from the ecdsa_pod
+        let mut main_pod_builder = MainPodBuilder::new(&params);
+        main_pod_builder.add_main_pod(main_ecdsa_pod.clone());
+
+        // add operation that ensures that the msg is as expected in the EcdsaPod
+        let msg_limbs = scalar_to_limbs(msg);
+        let msg_hash = PoseidonHash::hash_no_pad(&msg_limbs);
+        let msg_copy = main_pod_builder
+            .pub_op(op!(
+                new_entry,
+                (KEY_SIGNED_MSG, Value::from(RawValue(msg_hash.elements)))
+            ))
+            .unwrap();
+        main_pod_builder
+            .pub_op(op!(eq, (&main_ecdsa_pod, KEY_SIGNED_MSG), msg_copy))
+            .unwrap();
+
+        // perpetuate the pk
+        main_pod_builder
+            .pub_op(op!(copy, main_ecdsa_pod.public_statements[2].clone()))
+            .unwrap();
+
+        let mut prover = pod2::backends::plonky2::mock::mainpod::MockProver {};
+        let pod = main_pod_builder.prove(&mut prover, &params).unwrap();
+        assert!(pod.pod.verify().is_ok());
+
+        println!("going to prove the main_pod");
+        let mut prover = mainpod::Prover {};
+        let main_pod = timed!(
+            "main_pod_builder.prove",
+            main_pod_builder.prove(&mut prover, &params).unwrap()
+        );
+        let pod = (main_pod.pod as Box<dyn Any>)
+            .downcast::<mainpod::MainPod>()
+            .unwrap();
+        pod.verify().unwrap();
+
+        Ok(())
+    }
+}
```

Comment:
I would like to request an extra test that I think will be very useful for future introduction pods that will take a look at this implementation for reference:
The test would check that `pub_self_statements_target` and `pub_self_statements` match, and then it would calculate the id in circuit and natively and also check that they match.
I think this will help detecting issues on how the statements are generated and verified early on.

Reply:
Yes, that's a good point, I'll add it.